iOS_SDK/OneSignalSDK/Source/OSMessagingController.m

```diff
@@ -318,46 +325,61 @@ - (void)messageViewControllerWasDismissed {
-    
-    if (action.clickUrl)
-        [self handleMessageActionWithURL:action];
-    
-    if (self.actionClickBlock)
-        self.actionClickBlock(action);
-  
+- (void)sendClickRESTCall:(OSInAppMessage *)message withAction:(OSInAppMessageAction *)action {
     // Make sure no click tracking is performed for IAM previews
     // If the IAM clickId exists within the cached clickedClickIds return early so the click is not tracked
     // Handles body, button, or image clicks
     if (message.isPreview || [self.clickedClickIds containsObject:action.clickId])
         return;
-    
     // Add clickId to clickedClickIds
     [self.clickedClickIds addObject:action.clickId];
-    
+   
     let metricsRequest = [OSRequestInAppMessageClicked withAppId:OneSignal.app_id
                                                     withPlayerId:OneSignal.currentSubscriptionState.userId
                                                    withMessageId:message.messageId
                                                     forVariantId:message.variantId
                                                       withAction:action];
+   
+   [OneSignalClient.sharedClient executeRequest:metricsRequest
+                                      onSuccess:^(NSDictionary *result) {
+                                          NSString *successMessage = [NSString stringWithFormat:@"In App Message with id: %@, successful POST click update for click id: %@, with result: %@", message.messageId, action.clickId,  result];
+                                          [OneSignal onesignal_Log:ONE_S_LL_DEBUG message:successMessage];
+                                          
+                                          // Save the updated clickedClickIds since click was tracked successfully
+                                          [OneSignalUserDefaults.initStandard saveSetForKey:OS_IAM_CLICKED_SET_KEY withValue:self.clickedClickIds];
+                                      }
+                                      onFailure:^(NSError *error) {
+                                          NSString *errorMessage = [NSString stringWithFormat:@"In App Message with id: %@, failed POST click update for click id: %@, with error: %@", message.messageId, action.clickId, error];
+                                          [OneSignal onesignal_Log:ONE_S_LL_ERROR message:errorMessage];
+                                          
+                                          // Remove clickId from local clickedClickIds since click was not tracked
+                                          [self.clickedClickIds removeObject:action.clickId];
+                                      }];
+}
+
+- (void)sendOutcomeWithAction:(OSInAppMessageAction *)action {
+    if (action.outcome.unique) {
+        [OneSignal sendUniqueClickActionOutcome:action.outcome.name];
+    } else {
+       [OneSignal sendClickActionOutcomeWithValue:action.outcome.name value:action.outcome.weight > 0 ? action.outcome.weight : @0];
```

Comment:
We should check all 3 methods here instead of creating another layer.
For example:
```
if (action.outcome.unique) {
    [OneSignal sendUniqueOutcome:action.outcome.name];
} else if (action.outcome.weight > 0) {
    [OneSignal sendOutcomeWithValue:action.outcome.name value:action.outcome.weight];
} else {
    [OneSignal sendOutcome:action.outcome.name value:action.outcome.weight];
}
```